util/clustering/kmeans_clustering.py

```diff
@@ -0,0 +1,231 @@
+from sklearn.decomposition import PCA
+
+# for everything else
+import argparse
+import os
+import numpy as np
+import matplotlib.pyplot as plt
+from PIL import Image
+from tqdm import tqdm
+
+
+def main():
+    parser = argparse.ArgumentParser(description='Clustering')
+    if 'SM_OUTPUT_DATA_DIR' in os.environ: 
+
+        parser.add_argument('--output-data-dir', type=str, default=os.environ['SM_OUTPUT_DATA_DIR'])
+        #parser.add_argument('--model-dir', type=str, default=os.environ['SM_MODEL_DIR'])
+        parser.add_argument('--input-data-dir', type=str, default=os.environ['SM_MODEL_DIR'])
+    else: 
+        parser.add_argument('--output-data-dir', type=str)
+        parser.add_argument('--input-data-dir', type=str)   
+    
+    parser.add_argument('--analyze-clusters', type=int, default=3)   
+
+    args = parser.parse_args()
+
+    
+    file_dir = os.getcwd()
+    # # change the working directory to the path where the images are located
+    output_path = args.input_data_dir
+    os.chdir(output_path)
+
+    detection_train_images = []
+    # creates a ScandirIterator aliased as files
+    print('------------ getting image names ------------')
+    with os.scandir(output_path) as files:
+      # loops through each file in the directory
+        for file in files:
+            if file.name.endswith('.jpg'):
+              # adds only the image files to the detection_train_imagess list
+                detection_train_images.append(file.name)
+   
+    # print('get bucket info')
+    # bucket = "calvinandpogs-ee148"
+    # detection_train_images_path = "atrw/detection/train"
+    # s3 = boto3.resource("s3")
+    # s3_bucket = s3.Bucket(bucket)
+    # print('bucket info collected')
+
+    # # this list holds all the image filename
+    # detection_train_images = [f.key.split(detection_train_images_path + "/")[1] for f in s3_bucket.objects.filter(Prefix=detection_train_images_path).all()]
+    print(str(len(detection_train_images)) + ' images collected')
+
+        
+    model = VGG16()
+    model = Model(inputs = model.inputs, outputs = model.layers[-2].output)
+
+    def extract_features(file, model):
+        # load the image from the s3 bucket as a 224x224 array and convert from 
+        # 'PIL.Image.Image' to numpy array because the VGG model expects the images
+        # it receives to be 224x224 NumPy arrays.
+
+        # object = s3.Object(bucket, detection_train_images_path + '/' + file)
+        # img = load_img(BytesIO(object.get()['Body'].read()), target_size=(224,224))
+        img = load_img(file, target_size=(224,224))
+
+        img = np.array(img) 
+        # reshape the data for the model reshape(num_of_samples, dim 1, dim 2, channels)
+        reshaped_img = img.reshape(1,224,224,3) 
+        # prepare image for model
+        imgx = preprocess_input(reshaped_img)
+        # get the feature vector
+        features = model.predict(imgx, use_multiprocessing=True)
+        return features
+    
+    data = {}
```

Comment:
Bad variable name. 'data' is actually storing image features, not input data or another generic form